runtime/shibuya/src/lib.rs

```diff
@@ -1535,6 +1539,70 @@ impl pallet_migrations::Config for Runtime {
     type WeightInfo = pallet_migrations::weights::SubstrateWeight<Runtime>;
 }
 
+parameter_types! {
+    pub const EnterDuration: BlockNumber = 4 * HOURS;
+    pub const EnterDepositAmount: Option<Balance> = None;
+    pub const ExtendDuration: BlockNumber = 2 * HOURS;
+    pub const ExtendDepositAmount: Option<Balance> = None;
+    pub const ReleaseDelay: Option<u32> = None;
+}
+
+/// Calls that can bypass the safe-mode pallet.
+pub struct SafeModeWhitelistedCalls;
+impl Contains<RuntimeCall> for SafeModeWhitelistedCalls {
+    fn contains(call: &RuntimeCall) -> bool {
+        match call {
+            RuntimeCall::Sudo(_)
+            // System and Timestamp are required for block production
+            | RuntimeCall::System(_)
+            | RuntimeCall::Timestamp(_)
+            | RuntimeCall::SafeMode(_)
+            | RuntimeCall::TxPause(_) => true,
+            _ => false,
+        }
+    }
+}
+
+/// Calls that cannot be paused by the tx-pause pallet.
+pub struct TxPauseWhitelistedCalls;
+/// All calls can be paused, except Sudo calls.
+impl Contains<RuntimeCallNameOf<Runtime>> for TxPauseWhitelistedCalls {
+    fn contains(full_name: &pallet_tx_pause::RuntimeCallNameOf<Runtime>) -> bool {
+        matches!(full_name.0.as_slice(), b"Sudo")
+    }
+}
+
+impl pallet_safe_mode::Config for Runtime {
+    type RuntimeEvent = RuntimeEvent;
+    type Currency = Balances;
+    type RuntimeHoldReason = RuntimeHoldReason;
+    type WhitelistedCalls = SafeModeWhitelistedCalls;
+    type EnterDuration = EnterDuration;
+    type EnterDepositAmount = EnterDepositAmount;
+    type ExtendDuration = ExtendDuration;
+    type ExtendDepositAmount = ExtendDepositAmount;
+    // The 'Success' value below represents the number of blocks that the origin may induce safe mode
+    type ForceEnterOrigin =
+        EnsureWithSuccess<EnsureRootOrAllTechnicalCommittee, AccountId, EnterDuration>;
+    type ForceExtendOrigin =
+        EnsureWithSuccess<EnsureRootOrAllTechnicalCommittee, AccountId, ExtendDuration>;
+    type ForceExitOrigin = EnsureRootOrAllTechnicalCommittee;
+    type ForceDepositOrigin = EnsureRootOrAllTechnicalCommittee;
+    type ReleaseDelay = ReleaseDelay;
+    type Notify = ();
```

Comment:
Perhaps the `Notify` should notify dApp staking pallet about this development.

The `maintenance` mode there isn't just to block dApp staking calls, but to prevent further modifications of storage, progression of eras, etc.

Reply:
You're right, I was thinking of associating it with the `tx-pause` pallet somehow, but it's actually a more important security behavior as you mentioned.